core/src/main/java/hudson/TcpSlaveAgentListener.java

```diff
@@ -98,6 +98,10 @@ public TcpSlaveAgentListener(int port) throws IOException {
             throw (BindException)new BindException("Failed to listen on port "+port+" because it's already in use.").initCause(e);
         }
         this.configuredPort = port;
+        setUncaughtExceptionHandler((t, e) -> {
+            LOGGER.log(Level.SEVERE, "Uncaught exception in TcpSlaveAgentListener " + t, e);
+            shutdown();
```

Comment:
So, it stops the Listener, and it never tries to recover without the Jenkins restart? Then 🐛 

I propose to do the following:
* After shutdown schedule a delayed async task (e.g. 5 seconds), which starts a new thread.
* Modify the SEVERE message and say that the thread will be rescheduled
* Print INFO message when the listener thread starts up